### ISSUE TITLE
Custom Label Prediction LMv2

### ISSUE THREAD

**Issue by sheikhasim** [NielsRogge/Transformers-Tutorials#154]
Model i'm using : Layout LMv2 for token classification
issue:
I trained and fine-tuned the LMv2 model for custom token classification i.e. instead of labelling whole text present in image , only particular labels are generated using bounding box information generated by an external OCR.
I'm getting correct desired output for the test dataset.

How would i preapre an unseen image  (image not present in train or test dataset ) to submit to the model and get only the particular labels as output. 
Right now whole image is getting labelled.

**Comment by fraps12** [comment]
Not clearly question but anyway.

1. Feed to model only tokens that you want to predict
2. If you want to get only particular label named question(e.g. you have 4 labels in output - "question", "answer", "header", "other") - filter the output by label that token have. 
`if model_output_token[0] in ["question"]:
    do something`

**Comment by sheikhasim** [comment]
@fraps12 
This is the inference script that i'm using:
https://colab.research.google.com/drive/1lFs8g8OjangPDjGs0kS_mXn0IK2xC8nw

Can you please explain filtering outputs by label ?

**Comment by fraps12** [comment]
@sheikhasim I have no access to notebook you provided.

I can assume that in output from prediction you have something like that

`prediction_output = [['question', [0, 0, 0, 0], ''], ['question', [120, 83, 247, 99], 'Resultat'], ['question', [120, 117, 174, 125], 'NAMN']]`
where elements are
list[0] = label
list[1] = boundingBox
list[2] = Token word

So if you are interested only in tokens labeled with "question"
You can make something like that `only_question_pred = [output_token for output_token in prediction_output if output_token[0] in ["question"]]`

**Comment by sheikhasim** [comment]
@fraps12 
I have updated the access settings , please use the below link 
https://colab.research.google.com/drive/1lFs8g8OjangPDjGs0kS_mXn0IK2xC8nw?usp=sharing
sure, i'll try the above approach.
Thank you !

**Comment by sheikhasim** [comment]
@fraps12 
adding 'PROPERTY' label in the above code snippet , is labelling all the words marked as 'PROPERTY' in **true_predictions**.

in notebook , 

> ` print (true_predictions)
> output : ['PROPERTY', 'PROPERTY', 'PROPERTY', 'PROPERTY', 'PROPERTY', 'PREPENALTY', 'PROPERTY', 'PROPERTY', 'LOANAMOUNT'`....]


> `for prediction, box in zip(true_predictions, true_boxes):
>   predicted_label = prediction
>   if predicted_label in ["PROPERTY"]:
>       draw.rectangle(box, outline=label2color[predicted_label])
>       draw.text((box[0]+10, box[1]-10), text=predicted_label, 
>        fill=label2color[predicted_label], font=font)
> else:
> pass`

I want the model to label  the property address as the 'PROPERTY'  present in the input image.

**Comment by fraps12** [comment]
@sheikhasim So the issue is that some tokens have not that label that you expect?

**Comment by sheikhasim** [comment]
@fraps12 
yes , like the label should be "LOANTERM" but model is labelling it as "PROPERTY".
And the second issue is that all other non-desired tokens present in image , like like 'sale price' , 'borrower' etc. , shouldn't be labelled by the model

**Comment by fraps12** [comment]
If token is labeled with not desired label you should add more samples to your dataset. And look at number of epochs of training. In process of training look at loss you have. First set number of epochs to 100 and look when the loss is least, than it can grow. That is not good. Or add some logic to save best model based on loss.

For my task i added like 256 samples fully labeled. Labeling all tokens on doc is good practice and will improve the results

Second issue:

```python
    for prediction, box in zip(true_predictions, true_boxes):
        predicted_label = iob_to_label(prediction).lower()
        if predicted_label in ["question"]:
            draw.rectangle(box, outline=label2color[predicted_label])
            draw.text((box[0] + 10, box[1] - 10), text=predicted_label, fill=label2color[predicted_label], font=font)
    image.show()
```
will draw only desired labels on image

and
```python
    filtered_predictions = []
    for prediction, box in zip(true_predictions, true_boxes):
        predicted_label = iob_to_label(prediction).lower()
        if predicted_label in ["question"]:
            predicted_label.append([predicted_label, box])
    image.show()
```
will results in list of bbox and labels of tokens that you are interested in

**Comment by sheikhasim** [comment]
@fraps12 
Sure i'll try the mentioned approach.
Thanks for the help!!

**Comment by sheikhasim** [comment]
Hey @fraps12 
I filtered out the desired tokens and labels in **true_filtered_labels [ ]** list and bounding boxes in **true_filtered_bounding_boxes[ ]** list.
Now i want to extract the text from my input image using the **true_filtered_bounding_boxes [ ]** only i.e. Extracting the text from image using the above specific bounding box information.
I tried several approaches with easyocr , but it's not giving the desired text.
Do you have any suggestions on this ?

**Comment by fraps12** [comment]
@sheikhasim I guess you need to normalize bounding boxes that you get from easyOCR. While you get output from model from an image with 1000*1000 shape, your easyOCR i guess use image with other shape.

More convenient way is zip the normalized output from your OCR engine to list of lists in form of `[[(x1,y1,w1,h1), word_token1], [(x2,y2,w2,h2), word_token2]...]`
than you can check the boundingBoxes from your model output with bounding boxes from your OCR enging and zip the model output bounding boxes with text from ocr engine :)

something like:

```pyton
     filtered_predictions = []
    for prediction, box in zip(true_predictions, true_boxes):
        predicted_label = iob_to_label(prediction).lower()
        **text_value = ""**
        **if predicted_label in ["question"]:
               for word in true_words:
                    x = word["boundingBox"]["x"]
                    y = word["boundingBox"]["y"]
                    width = word["boundingBox"]["width"]
                    height = word["boundingBox"]["height"]

                    if  [x, y, width + x, height + y] == box:
                        text_value = word["value"]**
            filtered_predictions.append([predicted_label, box, **text_value**])
    image.show()



Actually that questions don't relate to transformers itself. Further ask questions smth like that on `stackoverflow` :)

**Comment by sheikhasim** [comment]
@fraps12 
The question is related to lmv2 not OCR though :)
After the model labels the input image , how to extract the labels as key-value pairs.
I think they have mentioned that in one of the issues and provided a notebook for that.
https://colab.research.google.com/drive/16wqA3oTUf7yzUKsSSZxiMf1443_ZO3wC
I'll check that once.

Thanks for the help !!
